selection/src/commonMain/kotlin/Order.kt

```diff
@@ -0,0 +1,15 @@
+package com.juul.krayon.selection
+
+import com.juul.krayon.element.Element
+
+public fun <E : Element, D> Selection<E, D>.order(): Selection<E, D> {
+    for (group in groups) {
+        val parent = group.parent ?: continue
+        for (node in group.nodes) {
+            if (node != null) {
+                parent.appendChild(parent.removeChild(node))
+            }
+        }
+    }
+    return this
+}
```

Comment:
I can't quite figure out what `order` does from the code. Does it reverse it?

Reply:
Sorry, this function probably needs the docs more than most. **tl;dr: `order` ensures that `Element`s in the element tree have the same order as the associated `data` lists, which will matter more when I implement more of the library.** Now, how does the element tree get out of sync with data? Good question.

When calling `.data` on a selection, D3 actually supports two modes: _index_-identified data, and _key_-identified data. Mode is entirely determined by the most recent call to `.data`, so it's possible to use _indexed_ for one call and _keyed_ for the next (but you probably shouldn't). Also, as far as I can tell, mode only matters when _changing_ data because the initial call to `.data` is going to be 100% `enter`, 0% `update`/`exit`. Right now, I've only implemented the _index_-identified mode (because it was easier). 

When dealing with _indexed_ data, identity is purely based on, well, index. This has a few side effects that really simplify the mental model:
- Length of the `data` list becomes the only factor when determining update/enter/exit selections. You will have `update` for any indices that exist in both lists, `enter` for any indices that exist in the new list only, and `exit` for any indices that exist in the old list only.
    - Enter will only add elements at the end of the list
    - Exit will only remove elements at the end of the list
- Elements will always match the order of `data` (making this `order` function kind of useless)

When dealing with _keyed_ data, none of the above bullets are guaranteed. Here's some sample code so I have some concrete stuff to go on:

```kotlin
data class Person(val name: String)

val root = RootElement()

fun onNewData(vararg people: Person) {
    root.asSelection()
        .selectAll(TextElement)
        // Note the hypothetical key lambda below, not yet implemented
        .data(people.toList()) { (data) -> data.name }
        .join(onEnter = { append { (data) -> TextElement(data.name) } })
}

// Insert data propagation method of choice
onNewData(Person("Alice"), Person("Bob"))
onNewData(Person("Charlie"), Person("Bob"))
onNewData(Person("Bob"), Person("Charlie"))
```

In _indexed_ mode (without the lambda argument in `data`), the second call to `onNewData` would have an `update` selection with two elements, and empty `enter`/`exit` selections. The third would have the same. Both calls would also have a bug, because text is only set when the element is first created, and not when it's updated. While this bug can often be worked around by doing more in `onUpdate` and less in `onEnter`, doing so often doesn't play nice with animations (you might want to show a person entering/leaving differently than a person changing their name). Because the element order always matches the data order, the `order` function does nothing.

In _keyed_ mode, the second call to `onNewData` is much more interesting: the `update` selection contains Bob, the `enter` selection contains Charlie, and the `exit` selection contains Alice. On the third call, `update` contains both Bob and Charlie, while `enter` and `exit` are empty. On this third call, the `order` function actually does something: even though no elements were added or removed, the element tree would have gotten out of sync with the data.

Reply:
Wow, thanks for the incredibly helpful/thorough explanation!